### PR TITLE
[FLINK-1376] [runtime] Add proper shared slot release in case of a fatal TaskManager failure for 0.8 branch

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR introduces SharedSlots as being a special Slot type and as such being released properly in case an Instance has been marked dead. This fixes the problem that a dead instance, which has not been shutdown properly, causes a job not being removed properly from the system, because it is not aware of the SubSlots.

Adds test cases where only the heartbeat thread of TaskManager is killed.

Except for the test cases, this is basically the same PR as #317 just rebased on the 0.8 branch.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/instance/SharedSlot.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+	/**
+	 * Disposes the given sub slot. This
+	 * is done by the means of the assignmentGroup in order to synchronize the method. If the
+	 * disposed slot was the last sub slot, then the shared slot is marked to be cancelled and is
+	 * disposed/returned to the owning instance.
+	 *
+	 * @param slot sub slot which shall be removed from the shared slot
+	 */
+	public void disposeChild(SimpleSlot slot){
```
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/instance/SharedSlot.java]: For a second I was wondering what a child is? Maybe we should just keep it as `disposeSubSlot` in order to be in sync with the javadocs?

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/instance/SharedSlot.java]: It used to be called disposeSubSlot but then I changed the slot system being a tree with root, parent and children. Can rename it if you think the term disposeSubSlot is easier to understand.

Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/instance/SharedSlot.java]: No, I think it's perfectly fine :)

**Comment by rmetzger** [comment]
I've merged it to the 0.8 branch. I forgot to close the PR, could you do it manually?
Thank you.

**Pull request closed without merging**